src/crimson/os/seastore/seastore_types.h

```diff
@@ -377,6 +377,162 @@ class segment_map_t {
+public:
+  block_id_t() = default;
+  constexpr block_id_t(device_id_t id, device_block_id_t block)
+    : block(make_internal(block, id)) {}
+
+  [[gnu::always_inline]]
+  device_id_t device_id() const {
+    return internal_to_device(block);
+  }
+
+  [[gnu::always_inline]]
+  constexpr device_block_id_t device_block_id() const {
+    return internal_to_block(block);
+  }
+
+  bool operator==(const block_id_t& other) const {
+    return block == other.block;
+  }
+  bool operator!=(const block_id_t& other) const {
+    return block != other.block;
+  }
+  bool operator<(const block_id_t& other) const {
+    return block < other.block;
+  }
+  bool operator<=(const block_id_t& other) const {
+    return block <= other.block;
+  }
+  bool operator>(const block_id_t& other) const {
+    return block > other.block;
+  }
+  bool operator>=(const block_id_t& other) const {
+    return block >= other.block;
+  }
+
+private:
+  static constexpr unsigned block_id_bits = (
+    BLK_ID_LEN_BITS
+  );
+
+  static inline device_id_t internal_to_device(internal_block_id_t id) {
+    return (static_cast<device_id_t>(id) & BLK_ID_MASK) >> block_id_bits;
+  }
+
+  constexpr static inline device_block_id_t internal_to_block(
+    internal_block_id_t id) {
+    return id & (~BLK_ID_MASK);
+  }
+
+  constexpr static inline internal_block_id_t make_internal(
+    device_block_id_t id,
+    device_id_t d_id) {
+    return static_cast<internal_block_id_t>(id) |
+      (static_cast<internal_block_id_t>(d_id) << block_id_bits) |
+      (static_cast<internal_block_id_t>(0x1) <<
+      (BLK_ID_LEN_BITS + DEVICE_ID_LEN_BITS - 1));
+  }
+
+  friend struct blk_paddr_t;
+  friend struct paddr_t;
+  friend struct paddr_le_t;
+};
+
+constexpr block_id_t NULL_BLK_ID = block_id_t(DEVICE_ID_NULL, 0);
+constexpr block_id_t FAKE_BLK_ID = block_id_t(DEVICE_ID_FAKE, 0);
+
+using device_block_off_t = int32_t;
+static constexpr uint16_t BLK_OFF_LEN_BITS = 20;
+constexpr device_block_off_t BLK_OFF_MAX = 
+  std::numeric_limits<device_block_off_t>::max() >>
+  (std::numeric_limits<device_block_off_t>::digits - BLK_OFF_LEN_BITS);
+constexpr device_block_off_t BLK_OFF_MIN =
+  std::numeric_limits<device_block_off_t>::min() /
+  std::pow(2, std::numeric_limits<device_block_off_t>::digits - BLK_OFF_LEN_BITS);
+
+struct block_off_t {
+  using internal_block_off_t = int32_t;
+  internal_block_off_t off = BLK_OFF_MAX;
+  constexpr void check_valid(const internal_block_off_t offset) const {
+    assert(offset < BLK_OFF_MAX);
+    assert(offset > BLK_OFF_MIN);
+  }
+  [[gnu::always_inline]]
+  constexpr device_block_off_t device_block_off() const {
+    return off;
+  }
+  constexpr block_off_t(device_block_off_t offset) : off(offset) {}
+  constexpr uint32_t get_raw_bits() const {
+    check_valid(off);
+    uint32_t ret = off & 0xFFFFF;
+    if (off < 0) {
+      ret |= 1 << BLK_OFF_LEN_BITS;
+    }
+    return ret;
+  }
+  bool operator==(const block_off_t& other) const {
```

Comment:
I don't think we need separate block_off_t and segment_off_t types -- how about we just rename segment_off_t seastore_off_t?